### PR TITLE
add support for branch name based testing, add simplified semver support

### PR DESCRIPTION
**Pull request by RonnyPfannschmidt**

### CODE REVIEW

File: testing/test_mercurial.py [hunk 1]
```diff
@@ -114,7 +114,7 @@ def test_parse_no_worktree(tmpdir):
 def version_1_0(wd):
     wd('hg branch default')
     wd.commit_testfile()
-    wd('hg tag 1.0 -u test -d "0 0"')
+    wd('hg tag 1.0.0 -u test -d "0 0"')
```
Comment (jaraco) [inline testing/test_mercurial.py:117]: What's happening here? Do mercurial tags now require full semver and drop support for N.N numbers?

File: testing/test_version.py [hunk 2]
```diff
@@ -0,0 +1,23 @@
+import pytest
+from setuptools_scm.version import meta, simplified_semver_version
+
+
+@pytest.mark.parametrize('version, expected_next', [
+    pytest.param(meta("1.0.0"), "1.0.0", id='exact'),
+
+    pytest.param(meta("1.0"), "1.0.0", id='short_tag'),
```
Comment (jaraco) [inline testing/test_version.py:8]: Does test imply that a tag of "1.0" generates a version of "1.0.0"? I'm unsure what the implications are of that. Is this an opt-in feature?

File: setuptools_scm/version.py [hunk 3]
```diff
@@ -145,6 +148,31 @@ def guess_next_dev_version(version):
...
+    if distance:
+        parts.append("dev" + str(distance))
+    return '.'.join(parts)
+
+
+def simplified_semver_version(version):
+    if version.exact:
+        return guess_next_simple_semver(version.tag, 0, 3, increment=False)
+    else:
+        if version.branch is not None and 'feature' in version.branch:
```
Comment (jaraco) [inline setuptools_scm/version.py]: Given that these two code branches are 90% the same, I'd write it as:
```
retain = 2 + (version.branch is not None and 'feature' in version.branch)
return guess_next_simple_semver(version.tag, version.distance, retain=retain)
```

File: setuptools_scm/version.py [hunk 4]
```diff
@@ -145,6 +148,31 @@ def guess_next_dev_version(version):
...
+    del parts[3:]
+    if distance:
+        parts.append("dev" + str(distance))
+    return '.'.join(parts)
+
+
+def simplified_semver_version(version):
+    if version.exact:
+        return guess_next_simple_semver(version.tag, 0, 3, increment=False)
+    else:
```
Comment (jaraco) [inline setuptools_scm/version.py]: This else is unneeded.

File: setuptools_scm/version.py [hunk 5]
```diff
@@ -145,6 +148,31 @@ def guess_next_dev_version(version):
         return guess_next_version(version.tag, version.distance)
 
 
+def guess_next_simple_semver(version, distance, retain, increment=True):
+    parts = str(version).split('.', retain)
```
Comment (jaraco) [inline setuptools_scm/version.py]: These two lines use retain twice for the same meaning. Consider replacing with
```
parts = str(version).split('.', retain)[:-1]
```

File: setuptools_scm/version.py [hunk 6]
```diff
@@ -145,6 +148,31 @@ def guess_next_dev_version(version):
...
+    del parts[retain:]
+    while len(parts) < retain:
```
Comment (jaraco) [inline setuptools_scm/version.py]: These next two lines are padding the list of version numbers with zeros. I'd either document that with a comment and whitespace or extract a function whose docstring described the intention.

File: setuptools_scm/version.py [hunk 7]
```diff
@@ -145,6 +148,31 @@ def guess_next_dev_version(version):
...
+        parts.append("0")
+    if increment:
+        parts[-1] = str(int(parts[-1]) + 1)
```
Comment (jaraco) [inline setuptools_scm/version.py]: It feels a little awkward here to be working with strings as numbers. It seems it'd be preferable to me to parse everything as numbers, and then convert back to a string at the end. But that may not be possible if some of the parts are meant to be non-numbers.

**Review by jaraco — APPROVED** [review]
The hardest part for me in reviewing this change is there's no documentation, so I'm not really sure what was intended. I only have one point of reference and that's "what changed". The topic of the PR suggests that two changes are happening here, "support for branch name based testing" and "add simplified semver support". Perhaps that suggests this should be two changes.

My biggest concern is about the apparent rewriting of N.N numbers to N.N.0, which may cause unexpected behavior and consternation for projects that intentionally release two-part version numbers. I'd like to know what is the story for such projects?

Otherwise, these changes seem pretty sane. I made a few comments about how I'd improve it, but overall, after spending 20-30 minutes on it, I think I understand what's going on and it seems fine.

File: testing/test_version.py (hunk 2 above)
Comment (RonnyPfannschmidt) [reply testing/test_version.py:8]: this is opt-in, one has to specifiy simplified semver

File: setuptools_scm/version.py (hunk 5 above)
Comment (RonnyPfannschmidt) [reply setuptools_scm/version.py]: @jaraco this needs a different expression perhaps, bur the `[:-1]` is wrong, as t fails for versions where the parts contain less than retain elements

File: setuptools_scm/version.py (hunk 7 above)
Comment (RonnyPfannschmidt) [reply setuptools_scm/version.py]: unfortunately its permissible for `parts[0]` to be `'v123'`

**Comment by gerritholl** [comment]
How do I use this?

**Comment by RonnyPfannschmidt** [comment]
@gerritholl by setting the version scheme to python-simplified-semver

I can elaborate once I get back to the computer later today

**Comment by wookayin** [comment]
pyproject.toml
```diff
 [tool.setuptools_scm]
 write_to = ...
+version_scheme = "python-simplified-semver"
```

The most important line is
```
if version.branch is not None and "feature" in version.branch:
```

so I think this version scheme bumps minor version when the branch name contains `feature`. (Not possible when the branch name is `main`, `master`, or `dev` ??)

**Pull request merged**